File: apps/server/internal/charging/backtest.go
package charging

import (
	"fmt"
	"math"
	"math/rand"
	"slices"
	"time"
)

type Scenario struct {
	ID                int     `json:"id"`
	StartIndex        int     `json:"startIndex"`
	EndIndexExclusive int     `json:"endIndexExclusive"`
	StartSOCPercent   float64 `json:"startSocPercent"`
	TargetSOCPercent  float64 `json:"targetSocPercent"`
}

type StrategyResult struct {
	Completed              bool    `json:"completed"`
	GridEnergyKWh          float64 `json:"gridEnergyKwh"`
	BatteryEnergyKWh       float64 `json:"batteryEnergyKwh"`
	EmissionsGCO2          float64 `json:"emissionsGco2"`
	AverageCarbonIntensity float64 `json:"averageCarbonIntensity"`
}

type ScenarioResult struct {
	Scenario                Scenario       `json:"scenario"`
	RequiredEnergyKWh       float64        `json:"requiredEnergyKwh"`
	AvailableHours          float64        `json:"availableHours"`
	Baseline                StrategyResult `json:"baseline"`
	Optimized               StrategyResult `json:"optimized"`
	EmissionsSavedGCO2      float64        `json:"emissionsSavedGco2"`
	EmissionsSavingsPercent float64        `json:"emissionsSavingsPercent"`
}

type BacktestSummary struct {
	Runs                    int     `json:"runs"`
	CompletedBaseline       int     `json:"completedBaseline"`
	CompletedOptimized      int     `json:"completedOptimized"`
	BaselineEmissionsGCO2   float64 `json:"baselineEmissionsGco2"`
	OptimizedEmissionsGCO2  float64 `json:"optimizedEmissionsGco2"`
	EmissionsSavedGCO2      float64 `json:"emissionsSavedGco2"`
	EmissionsSavingsPercent float64 `json:"emissionsSavingsPercent"`
	AverageSavingsPercent   float64 `json:"averageSavingsPercent"`
	MedianSavingsPercent    float64 `json:"medianSavingsPercent"`
	MinSavingsPercent       float64 `json:"minSavingsPercent"`
	MaxSavingsPercent       float64 `json:"maxSavingsPercent"`
}

type Backtest struct {
	Summary   BacktestSummary  `json:"summary"`
	Scenarios []ScenarioResult `json:"scenarios"`
}

func RunBacktest(points []CarbonPoint, vehicle Vehicle, runs int, seed int64) (Backtest, error) {
	if runs <= 0 || runs > 10_000 {
		return Backtest{}, fmt.Errorf("runs must be between 1 and 10000")
	}
	if err := validateVehicle(vehicle); err != nil {
		return Backtest{}, err
	}
	if len(points) < 25 {
		return Backtest{}, fmt.Errorf("at least 25 carbon forecast points are required")
	}
	points = slices.Clone(points)
	slices.SortFunc(points, func(a, b CarbonPoint) int { return a.Datetime.Compare(b.Datetime) })
	slotDuration := medianSlotDuration(points)
	if slotDuration <= 0 {
		return Backtest{}, fmt.Errorf("carbon forecast timestamps must increase")
	}

	random := rand.New(rand.NewSource(seed))
	minWindow := min(24, len(points))
	maxWindow := min(144, len(points))
	batteryPerSlot := vehicle.MaxChargePowerKW * vehicle.ChargingEfficiency * slotDuration.Hours()
	results := make([]ScenarioResult, 0, runs)

	for id := 1; id <= runs; id++ {
		windowSlots := randomIntInclusive(random, minWindow, maxWindow)
		startIndex := randomIntInclusive(random, 0, len(points)-windowSlots)
		startSOC := float64(randomIntInclusive(random, 10, 65))
		windowUtilization := 0.25 + random.Float64()*0.6
		desiredEnergy := float64(windowSlots) * batteryPerSlot * windowUtilization
		batteryHeadroom := vehicle.BatteryCapacityKWh * (90 - startSOC) / 100
		requiredEnergy := math.Min(desiredEnergy, batteryHeadroom)
		targetSOC := startSOC + requiredEnergy/vehicle.BatteryCapacityKWh*100

		scenario := Scenario{
			ID:                id,
			StartIndex:        startIndex,
			EndIndexExclusive: startIndex + windowSlots,
			StartSOCPercent:   startSOC,
			TargetSOCPercent:  targetSOC,
		}
		result := runBacktestScenario(points, vehicle, scenario, slotDuration)
		results = append(results, result)
	}

	return Backtest{Summary: summarize(results), Scenarios: results}, nil
}

func runBacktestScenario(
	points []CarbonPoint,
	vehicle Vehicle,
	scenario Scenario,
	slotDuration time.Duration,
) ScenarioResult {
	available := points[scenario.StartIndex:scenario.EndIndexExclusive]
	required := vehicle.BatteryCapacityKWh *
		(scenario.TargetSOCPercent - scenario.StartSOCPercent) / 100
	baselineOrder := make([]int, len(available))
	for index := range baselineOrder {
		baselineOrder[index] = index
	}
	optimizedOrder := slices.Clone(baselineOrder)
	slices.SortFunc(optimizedOrder, func(a, b int) int {
		if available[a].CarbonIntensity < available[b].CarbonIntensity {
			return -1
		}
		if available[a].CarbonIntensity > available[b].CarbonIntensity {
			return 1
		}
		return a - b
	})

	baseline := simulate(available, baselineOrder, vehicle, required, slotDuration)
	optimized := simulate(available, optimizedOrder, vehicle, required, slotDuration)
	saved := baseline.EmissionsGCO2 - optimized.EmissionsGCO2
	percent := 0.0
	if baseline.EmissionsGCO2 > 0 {
		percent = saved / baseline.EmissionsGCO2 * 100
	}
	return ScenarioResult{
		Scenario:                scenario,
		RequiredEnergyKWh:       required,
		AvailableHours:          float64(len(available)) * slotDuration.Hours(),
		Baseline:                baseline,
		Optimized:               optimized,
		EmissionsSavedGCO2:      saved,
		EmissionsSavingsPercent: percent,
	}
}

func simulate(
	points []CarbonPoint,
	order []int,
	vehicle Vehicle,
	requiredBatteryEnergy float64,
	slotDuration time.Duration,
) StrategyResult {
	remaining := requiredBatteryEnergy
	maxGridPerSlot := vehicle.MaxChargePowerKW * slotDuration.Hours()
	result := StrategyResult{}
	for _, index := range order {
		if remaining <= 1e-9 {
			break
		}
		gridEnergy := math.Min(maxGridPerSlot, remaining/vehicle.ChargingEfficiency)
		batteryEnergy := gridEnergy * vehicle.ChargingEfficiency
		result.GridEnergyKWh += gridEnergy
		result.BatteryEnergyKWh += batteryEnergy
		result.EmissionsGCO2 += gridEnergy * points[index].CarbonIntensity
		remaining -= batteryEnergy
	}
	result.Completed = remaining <= 1e-9
	if result.GridEnergyKWh > 0 {
		result.AverageCarbonIntensity = result.EmissionsGCO2 / result.GridEnergyKWh
	}
	return result
}

func summarize(results []ScenarioResult) BacktestSummary {
	summary := BacktestSummary{Runs: len(results)}
	savings := make([]float64, 0, len(results))
	for _, result := range results {
		if result.Baseline.Completed {
			summary.CompletedBaseline++
		}
		if result.Optimized.Completed {
			summary.CompletedOptimized++
		}
		summary.BaselineEmissionsGCO2 += result.Baseline.EmissionsGCO2
		summary.OptimizedEmissionsGCO2 += result.Optimized.EmissionsGCO2
		summary.AverageSavingsPercent += result.EmissionsSavingsPercent
		savings = append(savings, result.EmissionsSavingsPercent)
	}
	summary.EmissionsSavedGCO2 = summary.BaselineEmissionsGCO2 - summary.OptimizedEmissionsGCO2
	if summary.BaselineEmissionsGCO2 > 0 {
		summary.EmissionsSavingsPercent = summary.EmissionsSavedGCO2 / summary.BaselineEmissionsGCO2 * 100
	}
	if len(results) > 0 {
		summary.AverageSavingsPercent /= float64(len(results))
		slices.Sort(savings)
		summary.MinSavingsPercent = savings[0]
		summary.MaxSavingsPercent = savings[len(savings)-1]
		middle := len(savings) / 2
		if len(savings)%2 == 0 {
			summary.MedianSavingsPercent = (savings[middle-1] + savings[middle]) / 2
		} else {
			summary.MedianSavingsPercent = savings[middle]
		}
	}
	return summary
}

func validateVehicle(vehicle Vehicle) error {
	if vehicle.BatteryCapacityKWh <= 0 || vehicle.MaxChargePowerKW <= 0 {
		return fmt.Errorf("vehicle battery capacity and charging power must be positive")
	}
	if vehicle.ChargingEfficiency <= 0 || vehicle.ChargingEfficiency > 1 {
		return fmt.Errorf("charging efficiency must be in the range (0, 1]")
	}
	return nil
}

func randomIntInclusive(random *rand.Rand, minimum, maximum int) int {
	return minimum + random.Intn(maximum-minimum+1)
}

File: src/packages/charging/backtest.ts
import type { CarbonIntensity } from '../electricitymaps/api';

export interface BacktestVehicle {
  batteryCapacityKwh: number;
  maxChargePowerKw: number;
  chargingEfficiency: number;
}

export interface ChargingScenario {
  id: number;
  startIndex: number;
  endIndexExclusive: number;
  startSocPercent: number;
  targetSocPercent: number;
}

export interface SlotCharge {
  datetime: string;
  carbonIntensity: number;
  gridEnergyKwh: number;
  batteryEnergyKwh: number;
  emissionsGco2: number;
}

export interface StrategyResult {
  completed: boolean;
  deliveredBatteryEnergyKwh: number;
  gridEnergyKwh: number;
  emissionsGco2: number;
  averageCarbonIntensity: number;
  charges: SlotCharge[];
}

export interface ScenarioResult {
  scenario: ChargingScenario;
  requiredBatteryEnergyKwh: number;
  availableHours: number;
  baseline: StrategyResult;
  optimized: StrategyResult;
  emissionsSavedGco2: number;
  emissionsSavingsPercent: number;
}

export interface BacktestSummary {
  runs: number;
  completedBaseline: number;
  completedOptimized: number;
  baselineEmissionsGco2: number;
  optimizedEmissionsGco2: number;
  emissionsSavedGco2: number;
  emissionsSavingsPercent: number;
  averageSavingsPercent: number;
  medianSavingsPercent: number;
  minSavingsPercent: number;
  maxSavingsPercent: number;
}

const EPSILON = 1e-9;

export function inferSlotHours(points: CarbonIntensity[]): number {
  if (points.length < 2) throw new Error('At least two carbon-intensity points are required.');
  const differences = points
    .slice(1)
    .map((point, index) => Date.parse(point.datetime) - Date.parse(points[index].datetime))
    .filter((difference) => Number.isFinite(difference) && difference > 0)
    .toSorted((left, right) => left - right);
  if (!differences.length) throw new Error('Carbon-intensity timestamps must increase.');
  return differences[Math.floor(differences.length / 2)] / 3_600_000;
}

export function generateScenarios(
  pointCount: number,
  vehicle: BacktestVehicle,
  slotHours: number,
  runs = 100,
  seed = 20_260_812,
): ChargingScenario[] {
  if (pointCount < 25) throw new Error('At least 25 carbon-intensity points are required.');
  if (!Number.isInteger(runs) || runs <= 0) throw new Error('runs must be a positive integer.');

  const random = mulberry32(seed);
  const minWindowSlots = Math.min(24, pointCount);
  const maxWindowSlots = Math.min(144, pointCount);
  const batteryPerSlot = vehicle.maxChargePowerKw * vehicle.chargingEfficiency * slotHours;
  const scenarios: ChargingScenario[] = [];

  for (let id = 1; id <= runs; id += 1) {
    const windowSlots = randomInteger(random, minWindowSlots, maxWindowSlots);
    const startIndex = randomInteger(random, 0, pointCount - windowSlots);
    const startSocPercent = randomInteger(random, 10, 65);
    const windowUtilization = 0.25 + random() * 0.6;
    const desiredEnergy = windowSlots * batteryPerSlot * windowUtilization;
    const batteryHeadroom = (vehicle.batteryCapacityKwh * (90 - startSocPercent)) / 100;
    const requiredEnergy = Math.min(desiredEnergy, batteryHeadroom);
    const targetSocPercent = startSocPercent + (requiredEnergy / vehicle.batteryCapacityKwh) * 100;

    scenarios.push({
      id,
      startIndex,
      endIndexExclusive: startIndex + windowSlots,
      startSocPercent,
      targetSocPercent,
    });
  }

  return scenarios;
}

export function runScenario(
  points: CarbonIntensity[],
  vehicle: BacktestVehicle,
  scenario: ChargingScenario,
  slotHours = inferSlotHours(points),
): ScenarioResult {
  validateVehicle(vehicle);
  const available = points.slice(scenario.startIndex, scenario.endIndexExclusive);
  if (!available.length) throw new Error(`Scenario ${scenario.id} has no available slots.`);

  const requiredBatteryEnergyKwh =
    (vehicle.batteryCapacityKwh * (scenario.targetSocPercent - scenario.startSocPercent)) / 100;
  if (requiredBatteryEnergyKwh <= 0) {
    throw new Error(`Scenario ${scenario.id} target SOC must exceed start SOC.`);
  }

  // Baseline: start immediately and charge continuously.
  const baseline = simulateStrategy(
    available.map((_, index) => index),
    available,
    vehicle,
    requiredBatteryEnergyKwh,
    slotHours,
  );

  // Optimized formula: allocate energy to the lowest-carbon slots first.
  const optimizedOrder = available
    .map((point, index) => ({ index, carbonIntensity: point.carbonIntensity }))
    .toSorted(
      (left, right) => left.carbonIntensity - right.carbonIntensity || left.index - right.index,
    )
    .map(({ index }) => index);
  const optimized = simulateStrategy(
    optimizedOrder,
    available,
    vehicle,
    requiredBatteryEnergyKwh,
    slotHours,
  );

  const emissionsSavedGco2 = baseline.emissionsGco2 - optimized.emissionsGco2;
  const emissionsSavingsPercent =
    baseline.emissionsGco2 > 0 ? (emissionsSavedGco2 / baseline.emissionsGco2) * 100 : 0;

  return {
    scenario,
    requiredBatteryEnergyKwh,
    availableHours: available.length * slotHours,
    baseline,
    optimized,
    emissionsSavedGco2,
    emissionsSavingsPercent,
  };
}

export function summarizeBacktest(results: ScenarioResult[]): BacktestSummary {
  if (!results.length) throw new Error('At least one scenario result is required.');
  const savings = results
    .map((result) => result.emissionsSavingsPercent)
    .toSorted((left, right) => left - right);
  const baselineEmissionsGco2 = sum(results.map((result) => result.baseline.emissionsGco2));
  const optimizedEmissionsGco2 = sum(results.map((result) => result.optimized.emissionsGco2));
  const emissionsSavedGco2 = baselineEmissionsGco2 - optimizedEmissionsGco2;

  return {
    runs: results.length,
    completedBaseline: results.filter((result) => result.baseline.completed).length,
    completedOptimized: results.filter((result) => result.optimized.completed).length,
    baselineEmissionsGco2,
    optimizedEmissionsGco2,
    emissionsSavedGco2,
    emissionsSavingsPercent:
      baselineEmissionsGco2 > 0 ? (emissionsSavedGco2 / baselineEmissionsGco2) * 100 : 0,
    averageSavingsPercent: sum(savings) / savings.length,
    medianSavingsPercent: percentile(savings, 0.5),
    minSavingsPercent: savings[0],
    maxSavingsPercent: savings[savings.length - 1],
  };
}

function simulateStrategy(
  slotOrder: number[],
  points: CarbonIntensity[],
  vehicle: BacktestVehicle,
  requiredBatteryEnergyKwh: number,
  slotHours: number,
): StrategyResult {
  const maxGridEnergyPerSlot = vehicle.maxChargePowerKw * slotHours;
  let remainingBatteryEnergy = requiredBatteryEnergyKwh;
  const charges: SlotCharge[] = [];

  for (const index of slotOrder) {
    if (remainingBatteryEnergy <= EPSILON) break;
    const point = points[index];
    const requiredGridEnergy = remainingBatteryEnergy / vehicle.chargingEfficiency;
    const gridEnergyKwh = Math.min(maxGridEnergyPerSlot, requiredGridEnergy);
    const batteryEnergyKwh = gridEnergyKwh * vehicle.chargingEfficiency;
    charges.push({
      datetime: point.datetime,
      carbonIntensity: point.carbonIntensity,
      gridEnergyKwh,
      batteryEnergyKwh,
      emissionsGco2: point.carbonIntensity * gridEnergyKwh,
    });
    remainingBatteryEnergy -= batteryEnergyKwh;
  }

  const deliveredBatteryEnergyKwh = sum(charges.map((charge) => charge.batteryEnergyKwh));
  const gridEnergyKwh = sum(charges.map((charge) => charge.gridEnergyKwh));
  const emissionsGco2 = sum(charges.map((charge) => charge.emissionsGco2));

  return {
    completed: remainingBatteryEnergy <= EPSILON,
    deliveredBatteryEnergyKwh,
    gridEnergyKwh,
    emissionsGco2,
    averageCarbonIntensity: gridEnergyKwh > 0 ? emissionsGco2 / gridEnergyKwh : 0,
    charges: charges.toSorted(
      (left, right) => Date.parse(left.datetime) - Date.parse(right.datetime),
    ),
  };
}

function validateVehicle(vehicle: BacktestVehicle) {
  if (vehicle.batteryCapacityKwh <= 0 || vehicle.maxChargePowerKw <= 0) {
    throw new Error('Vehicle capacity and charging power must be positive.');
  }
  if (vehicle.chargingEfficiency <= 0 || vehicle.chargingEfficiency > 1) {
    throw new Error('Vehicle charging efficiency must be in the range (0, 1].');
  }
}

function mulberry32(seed: number) {
  let value = seed >>> 0;
  return () => {
    value += 0x6d2b79f5;
    let result = value;
    result = Math.imul(result ^ (result >>> 15), result | 1);
    result ^= result + Math.imul(result ^ (result >>> 7), result | 61);
    return ((result ^ (result >>> 14)) >>> 0) / 4_294_967_296;
  };
}

function randomInteger(random: () => number, min: number, max: number) {
  return Math.floor(random() * (max - min + 1)) + min;
}

function sum(values: number[]) {
  return values.reduce((total, value) => total + value, 0);
}

function percentile(sortedValues: number[], ratio: number) {
  const index = (sortedValues.length - 1) * ratio;
  const lower = Math.floor(index);
  const upper = Math.ceil(index);
  if (lower === upper) return sortedValues[lower];
  return sortedValues[lower] + (sortedValues[upper] - sortedValues[lower]) * (index - lower);
}
